ec2-spot-sagemaker-managed-spot-training/sagemaker-built-in-object-detection/tools/birdsOnEdge.py

```diff
@@ -0,0 +1,225 @@
+# *****************************************************
+#                                                    *
+# Copyright 2019 Amazon.com, Inc. or its affiliates. *
+# All Rights Reserved.                               *
+#                                                    *
+# *****************************************************
+import datetime
+import json
+import os
+import sys
+import time
+import urllib
+import zipfile
+from threading import Event, Thread, Timer
+
+import awscam
+import cv2
+import greengrasssdk
+import mo
+import numpy as np
+
+# Create a greengrass core sdk client
+client = greengrasssdk.client("iot-data")
+
+# The information exchanged between IoT and clould has a topic and a
+# message body. This is the topic used to send messages to cloud.
+iot_topic = "$aws/things/{}/infer".format(os.environ["AWS_IOT_THING_NAME"])
+
+client.publish(topic=iot_topic, payload="At start of lambda function")
+
+# boto3 is not installed on device by default. Install it if it is not
+# already present.
+
+boto_dir = "/tmp/boto_dir"
+if not os.path.exists(boto_dir):
+    os.mkdir(boto_dir)
+    urllib.urlretrieve("https://s3.amazonaws.com/dear-demo/boto_3_dist.zip", "/tmp/boto_3_dist.zip")
+    client.publish(topic=iot_topic, payload="Extracting boto3 distribution...")
+    with zipfile.ZipFile("/tmp/boto_3_dist.zip", "r") as zip_ref:
+        zip_ref.extractall(boto_dir)
+client.publish(topic=iot_topic, payload="Adding boto to path...")
+sys.path.append(boto_dir)
+
+import boto3
+
+client.publish(topic=iot_topic, payload="Completed import of boto3")
+
+
+class LocalDisplay(Thread):
+    """Class for facilitating the local display of inference results
+    (as images). The class is designed to run on its own thread. In
+    particular the class dumps the inference results into a FIFO
+    located in the tmp directory (which lambda has access to). The
+    results can be rendered using mplayer by typing:
+    mplayer -demuxer lavf -lavfdopts format=mjpeg:probesize=32 /tmp/results.mjpeg
+    """
+
+    def __init__(self, resolution):
+        """resolution - Desired resolution of the project stream"""
+        # Initialize the base class, so that the object can run on its own
+        # thread.
+        super(LocalDisplay, self).__init__()
+        # List of valid resolutions
+        RESOLUTION = {"1080p": (1920, 1080), "720p": (1280, 720), "480p": (858, 480)}
+        if resolution not in RESOLUTION:
+            raise Exception("Invalid resolution")
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.

Reply:
I'm on the fence on this one. While I completely agree with the snippet in production code, the context of this notebook is training and education focused on Spot adoption. Exception handling is quite specific on each system we write, adding a custom exception in this case will not change things on how we treat or handle the exception. Creating extra custom exceptions to treat each individual error type in this case, may deviate some of the attention from the actual exercise. Happy to get to a compromise and add a comment line that points to the importance of exception writing strategies. 

Would love to hear other feedback.